File: leetcode/python/0141_linked_list_cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head == None: return False
        if head.next == None: return False
        slow = head
        fast = head.next

        while slow != fast:
            if slow == None or fast == None: return False
            slow = slow.next
            fast = fast.next
            if fast == None: return False
            else: fast = fast.next

        return True

File: leetcode/java/0141_linked_list_cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) return false;

        ListNode hare = head;
        do {
            if (hare == null) return false;
            head = head.next;
            hare = hare.next;
            if (hare == null) {
                return false;
            } else {
                hare = hare.next;
            }
        } while (head != hare);

        if (hare == null || head == null) return false;

        return true;
    }
}